src/mpid/common/genq/mpidu_genq_shmem_pool.c

```diff
@@ -54,13 +54,14 @@ static int cell_block_alloc(MPIDU_genqi_shmem_pool_s * pool, int block_idx)
 
     new_cell_headers =
         (MPIDU_genqi_shmem_cell_header_s **) MPL_malloc(pool->cells_per_free_queue
+                                                        * pool->num_free_queue
                                                         * sizeof(MPIDU_genqi_shmem_cell_header_s *),
                                                         MPL_MEM_OTHER);
     MPIR_ERR_CHKANDJUMP(!new_cell_headers, rc, MPI_ERR_OTHER, "**nomem");
     pool->cell_headers = new_cell_headers;
 
     /* init cell headers */
-    int cells_per_block = pool->num_free_queue;
+    int cells_per_block = pool->cells_per_free_queue;
```

Comment:
Wait. you only allocated 1 or 2 cells per rank, and the performance on Intel was unaffected? Was that due to Intel's atomic performance was too slow to matter?

Reply:
With `MPIR_CVAR_CH4_SHM_POSIX_TOPO_ENABLE=0`, you are using a single cell per-rank, right? So when sending a large message, it is doing lock-step copy-in/copy-out on small chunks. Are you sure you don't see performance differences on Intel architectures?

Reply:
That's right. I was equally surprised. The following are `osu_bw` before and after the topo patch (without the fix in this PR).

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{background:#E2EFDA;
	mso-pattern:black none;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Before TOPO | TOPO_ENABLE=0
-- | --
4.47 | 4.45
9.06 | 8.73
18.09 | 17.84
36.05 | 35.68
72.09 | 71.08
143.81 | 142.21
237.96 | 240.56
483.54 | 471.39
897.22 | 897.68
1663.38 | 1692.81
3160.99 | 2966.03
4192.24 | 4403.79
5092.73 | 5273.76
6454.57 | 6759.22
6105.7 | 6377.65
6413.67 | 6573
6546.41 | 6856.04
6705.65 | 7052.65
6819.39 | 7201.67
7030.93 | 7393.26
8073.07 | 8279.48
7328.59 | 7022.39
7244.68 | 5989.3
7218.74 | 6111
7187.1 | 6509.18
7196.72 | 7239.55
6182.98 | 6206.13
5533.84 | 5611.83



</body>

</html>

Reply:
It looks like there is no overlapping between sender side copy and receiver side copy, since the single buffer case is getting similar performance. I suspect that it is not the atomics are slow---I imagine that single buffer mode would be pretty bad because it would need multiple atomic ops to do one buffer copy. I haven't figured out why the performance on Intel was subpar.

Reply:
Do you have the AMD numbers?

Reply:
AMD numbers

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/yguo/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{background:#E2EFDA;
	mso-pattern:black none;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">


Msg Size | Before TOPO | TOPO_ENABLE=0
-- | -- | --
1 | 10.03 | 3.56
2 | 20.46 | 7.69
4 | 36.73 | 14.81
8 | 82.99 | 30.8
16 | 163.79 | 61.12
32 | 328.33 | 123.06
64 | 655.75 | 229.88
128 | 1251.08 | 446.43
256 | 2407.35 | 877.05
512 | 3852.29 | 1645.49
1024 | 7560.24 | 2535.51
2048 | 13236.79 | 4943.92
4096 | 20123.48 | 8580.93
8192 | 27441.57 | 11615.52
16384 | 22934.71 | 9553.93
32768 | 30239.98 | 11793.14
65536 | 34247.95 | 13057.41
131072 | 37003.09 | 14189.15
262144 | 36298.74 | 14692.71
524288 | 31300.72 | 14409.15
1048576 | 27742.9 | 14292.61
2097152 | 27672.01 | 13384.58
4194304 | 38894.96 | 14289.3



</body>

</html>

Reply:
Look at Intel's miserable numbers! Maybe there is something in Intel's design that prevents concurrent copy-in/copy-out by two different processes within the same page?

@yfguo Could you check the AMD number after this PR. Let's confirm we fixed the issue before merging this PR.

Reply:
@hzhou Added one more column for the numbers of this PR. I am not sure what is the limitation on Intel side, still investigating.